pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/LargeClusterRoutingTableBuilder.java

```diff
@@ -0,0 +1,414 @@
+    // segments served does not cover all of the segments, we compute the list of missing segments and pick a random
+    // server that serves these missing segments until we have complete coverage of all the segments.
+    //
+    // We then order the segments in ascending number of replicas within our server set, in order to allocate the
+    // segments with fewer replicas first. This ensures that segments that are 'easier' to allocate are more likely to
+    // end up on a replica with fewer segments.
+    //
+    // Then, we pick a random replica for each segment, iterating from fewest replicas to most replicas, inversely
+    // weighted by the number of segments already assigned to that replica.
+    //
+    // Since there is a small probability of any given routing table ends up with many more segments to a single
+    // instance, we then filter out the routing tables based on a workload disparity metric (max number of segments
+    // assigned to a replica - min number of segments assigned to a replica).
+    //
+    // The algorithm is thus:
+    // 1. Compute the inverse external view, a mapping of servers to segments
+    // 2. For each server:
+    //   a) Pick (maxServersPerQuery - 1) additional distinct servers
+    //   b) Check if the server set covers all the segments; if not, add additional servers until it does.
+    //   c) Order the segments in our server set in ascending order of number of replicas present in our server set
+    //   d) For each segment, pick a random replica with proper weighting
+    //   e) Return that routing table
+    // 3. Filter the bottom 20% of the routing tables when ordered by the workload disparity metric
+
+    Random random = new Random();
+    long startTime = System.currentTimeMillis();
+
+    // TODO jfim: Add instance pruning
+
+    // Compute the inverse of the external view
+    Map<String, Set<String>> segmentToInstanceMap = new HashMap<>();
+    Map<String, Set<String>> instanceToSegmentMap = new HashMap<>();
+    Set<String> segmentsWithAtLeastOneOnlineReplica = new HashSet<>();
+    for (String partition : externalView.getPartitionSet()) {
+      Set<String> instancesForSegment = new HashSet<>();
+      segmentToInstanceMap.put(partition, instancesForSegment);
+
+      for (Map.Entry<String, String> instanceAndState : externalView.getStateMap(partition).entrySet()) {
+        String instance = instanceAndState.getKey();
+        String state = instanceAndState.getValue();
+
+        // Only consider partitions that are ONLINE
+        if (!"ONLINE".equals(state)) {
+          continue;
+        }
+
+        // Add to the instance -> segments map
+        Set<String> partitionsForInstance = instanceToSegmentMap.get(instance);
+
+        if (partitionsForInstance ==  null) {
+          partitionsForInstance = new HashSet<>();
+          instanceToSegmentMap.put(instance, partitionsForInstance);
+        }
+
+        partitionsForInstance.add(partition);
```

Comment:
suggest rename partition to segmentId
and maybe rename the set variables as well